### DISCUSSION TITLE
Building on windows w/ GCC10 version of ARM Toolchain (latest)

### DISCUSSION THREAD

**Discussion by damodaran-ai** [pybricks/support#194] [General]
H pybricks team!,

Firstly: thanks for all the efforts on pybricks! It is truly great! 

TL;DR - If you need to build with GCC10 it seems you need to add **Wno-error=zero-length-bounds** and **-fcommon** in the **stm32.mk** file. 

This is a very non urgent observation on toolchain behavior when following the instructions to build pybricks-micropython.  I was setting up a development environment to putter around with the IMU in experimental. Since I couldn't get the BLE communication working with InventorHub on OSX (different issue),  I switched to Windows and setup a dev environment following the instructions on the Contributions page. I tried building with the latest Arm Toolchain build however it failed with a -Werror=zero-length-bounds due to the variable length struct in the lpf2 module. I worked past that by adding a **-Wno-error=zero-length-bounds** in the stm32.mk file. Now it only prints the warning :

```
../../lib/pbio/drv/ioport/ioport_lpf2.c: In function 'process_thread_pbdrv_ioport_lpf2_process':
../../lib/pbio/drv/ioport/ioport_lpf2.c:520:74: warning: array subscript 0 is outside the bounds of an interior zero-length array 'const pbio_iodev_mode_t[0]' [-Wzero-length-bounds]
  520 |                         const lump_mode_flags_t *flags = &info->mode_info[0].flags;
      |                                                           ~~~~~~~~~~~~~~~^~~
In file included from ../../lib/pbio/drv/ioport/ioport_lpf2.c:19:
../../lib/pbio/include/pbio/iodev.h:349:23: note: while referencing 'mode_info'
  349 |     pbio_iodev_mode_t mode_info[0];
      |                       ^~~~~~~~~
```
It then failed on the linker with almost every .o file having many duplicate definitions due to common.h (some sample messages below):
```
arm-none-eabi-ld.exe: build/pybricks/common/pb_type_colorlight_external.o:(.rodata.pb_type_Speaker+0x0): multiple definition of `pb_type_Speaker'; build/pybricks/common/pb_type_battery.o:(.rodata.pb_type_Speaker+0x0): first defined here
arm-none-eabi-ld.exe: build/pybricks/common/pb_type_colorlight_external.o:(.rodata.pb_type_DCMotor+0x0): multiple definition of `pb_type_DCMotor'; build/pybricks/common/pb_type_battery.o:(.rodata.pb_type_DCMotor+0x0): first defined here
```

It appears GCC10 makes -fno-common to be the default instead of **-fcommon** so after adding that to the CFLAGS as well, 
voila it worked!

Just figured I'd post this out in case anyone else also tried with GCC10 and ran into the same issues. Apologies if this is obvious.

**Comment by laurensvalk** [comment]
It's cool to see that you are building the code :+1: This may indeed be worth resolving at some point.

> I was setting up a development environment to putter around with the IMU in experimental.

If you update to the latest build, you'll find some things in InventorHub.imu already. We're in the process of making it non-experimental.

I'm curious, did you build from source to tweak some of the IMU device settings? If so, which ones? We're contemplating which settings should be user-accessible from the Python script. So if you need anything in particular, now is a good time to ask :) 

Also check out [this branch](https://docs.pybricks.com/en/work/) of the docs, which has some of it documented as well.

**Comment by dlech** [comment]
We will also gladly take pull requests to fix the errors. Although I would rather fix the problem rather than ignore the errors.

**Reply by damodaran-ai** [reply]
Will definitely do next time! Thanks for the help.

**Comment by dlech** [comment]
Fixed in https://github.com/pybricks/pybricks-micropython/commit/14654bbaba54aa18e48c34a81b42e8584cd3d24a and https://github.com/pybricks/pybricks-micropython/commit/220afa2f55d52ae049bd77329e38ff1572777f0a

**Comment by dlech** [comment]
I also tried GCC 10 on our other hubs, but they wouldn't boot using firmware compiled with GCC 10. So apparently something is broken in GCC 10.

**Reply by dlech** [reply]
Fixed in https://github.com/pybricks/pybricks-micropython/commit/2219d726009b023216191a525ac54530eefce4f2
